transport/controller.cc

```diff
@@ -332,14 +332,22 @@ future<> controller::subscribe_server(sharded<cql_server>& server) {
     return server.invoke_on_all([this] (cql_server& server) {
         _mnotifier.local().register_listener(server.get_migration_listener());
         _lifecycle_notifier.local().register_subscriber(server.get_lifecycle_listener());
+        if (!_used_by_maintenance_socket) {
+            _sl_controller.local().register_subscriber(server.get_qos_configuration_listener());
+        }
         return make_ready_future<>();
     });
 }
 
 future<> controller::unsubscribe_server(sharded<cql_server>& server) {
     return server.invoke_on_all([this] (cql_server& server) {
         return _mnotifier.local().unregister_listener(server.get_migration_listener()).then([this, &server]{
-            return _lifecycle_notifier.local().unregister_subscriber(server.get_lifecycle_listener());
+            return _lifecycle_notifier.local().unregister_subscriber(server.get_lifecycle_listener()).then([this, &server] {
+                if (!_used_by_maintenance_socket) {
+                    return _sl_controller.local().unregister_subscriber(server.get_qos_configuration_listener());
+                }
+                return make_ready_future<>();
+            });
         });
     });
```

Comment:
Looks like a good opportunity to coroutinize this function (in a separate commit that would precede this commit)